### ISSUE TITLE
[F#] Implement new Concept Exercise: exceptions

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#981]
This issue describes how to implement the `exceptions` concept exercise for the F# track.

## Goal

The goal of this exercise is to teach the student the basics of the Concept of [Exceptions in F#][exception-handling].

## Learning objectives

- Know what exceptions are.
- Know when an exception should be raised.
- Know how to raise (and reraise) an exception.
- Know how to catch an exception.
- Know how to use a finally block to always execute logic even when there are exceptions.
- Know the most important built-in exceptions (`ArgumentException`, `InvalidOperationException`) and built-in functions to raise them.
- Know how to define a custom exception.
- Know how to use pattern matching with exceptions.
- Know when to use exceptions vs results.

## Out of scope

- Memory and performance characteristics.

## Concepts

The Concepts this exercise unlocks are:

- `exceptions-basic`: know what exceptions are; know when an exception should be raised; know how to raise (and reraise) an exception; know how to catch an exception; know how to use a finally block to always execute logic even when there are exceptions; know the most important built-in exceptions (`ArgumentException`, `InvalidOperationException`) and built-in functions to raise them; know how to define a custom exception; know how to use pattern matching with exceptions; know when to use exceptions vs results.
- `pattern-matching-types`: know how to pattern match on a specific type.

## Prerequisites

This exercise's prerequisites Concepts are:

- `pattern-matching-basic`: know how to do pattern matching.
- `strings-basic`: know how to do string formatting (used in the `failwithf` function).
- `inheritance`: know how to inherit from a class.

### Hints

- [Exception handling][exception-handling]: exception handling in F#.
- [Exception introduction][exeptions]: introduction to exceptions in F#.

### After

- [Exception handling][exception-handling]: exception handling in F#.
- [Exception introduction][exeptions]: introduction to exceptions in F#.
- [Exceptions vs Result][exceptions-vs-result]: shows the difference between using exceptions vs Results.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific logic to be added to the [analyzer][analyzer].

## Implementing

To implement the concept exercise, the following files must be created:

<pre>
languages
└── fsharp
    └── exercises
        └── concept
            └── exceptions
                ├── .docs
                |   ├── instructions.md
                |   ├── introduction.md
                |   ├── hints.md
                |   └── after.md
                ├── .meta
                |   ├── config.json
                |   |── design.md
                |   └── Example.fs
                ├── Exceptions.fs
                ├── Exceptions.fsproj
                └── ExceptionsTests.fs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should be explicit about what the student should learn from the exercise, and provide a short, concise introduction to the concept(s). The aim is to give the student just enough context to figure things out themselves and solve the exercise, as research has shown that self-discovery is the most effective learning experience. Mentioning technical terms that the student can Google if they so want, is preferable over including any code samples or an extensive description. For example we might describe a string as a "Sequence of Unicode characters" or a "series of bytes" or "an object". Unless the student needs to understand the details of what those mean to be able to solve the exercise we should not give more info in this introduction - instead allowing the student to Google, ignore, or map their existing knowledge.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

These files are also all described in the [concept exercises document][docs-concept-exercises].

## Step 5: update languages/fsharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "exceptions",
        "uuid": "1aa9ea53-ace0-41e4-bf1f-f4c2933c445b",
        "concepts": ["exceptions-basic", "pattern-matching-types"],
        "prerequisites": [
          "pattern-matching-basic",
          "strings-basic",
          "inheritance"
        ]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the F# track:

- `Exceptions.fs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `Exceptions.fsproj`: the F# project file.
- `ExceptionsTests.fs`: the test suite.
- `.meta/Example.fs`: an example implementation that passes all the tests. The project file should _not_ include this file.

## Step 7: update the general concept document

Add the exercise to the concept's shared document's `## Implementations` section ([example](https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations)). Note that there currently is no such document, so it has to be created first.

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][meta-design]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][meta-config.json]).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/fsharp/reference/README.md#reference-docs
[analyzer]: https://github.com/exercism/fsharp-analyzer
[representer]: https://github.com/exercism/fsharp-representer
[meta-design]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/design.md
[meta-config.json]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/config.json
[docs-concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[exeptions]: https://fsharpforfunandprofit.com/posts/exceptions/
[exception-handling]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/exception-handling/
[exceptions-vs-result]: https://danielwertheim.se/exceptions-vs-the-result-type-in-fsharp/
